src/core/utils/unique_service_manager.cpp

```diff
@@ -0,0 +1,299 @@
+                                                             bool &           aRlocIn)
+{
+    otError               error    = OT_ERROR_NONE;
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+    ThreadNetif &         netif    = GetNetif();
+    ServiceEntry *        entry    = &mEntries[aEntryIndex];
+    otServiceConfig       config;
+    bool                  serviceExistsInNetworkDataLeader = false;
+
+    aRlocIn = false;
+
+    while (netif.GetNetworkDataLeader().GetNextService(&iterator, &config) == OT_ERROR_NONE)
+    {
+        if (entry->mServiceConfig->mEnterpriseNumber != config.mEnterpriseNumber ||
+            entry->mServiceConfig->mServiceDataLength != config.mServiceDataLength ||
+            memcmp(entry->mServiceConfig->mServiceData, config.mServiceData, config.mServiceDataLength) != 0)
+        {
+            continue;
+        }
+
+        if (!serviceExistsInNetworkDataLeader)
+        {
+            serviceExistsInNetworkDataLeader = true;
+            aConfig                          = config;
+        }
+        else if (!entry->mServerCompareCallback(&aConfig.mServerConfig, &config.mServerConfig, entry->mContext))
+        {
+            aConfig = config;
+        }
+
+        if (config.mServerConfig.mRloc16 == GetNetif().GetMle().GetRloc16())
+        {
+            aRlocIn = true;
+        }
+    }
+
+    VerifyOrExit(serviceExistsInNetworkDataLeader, error = OT_ERROR_NOT_FOUND);
+
+exit:
+    return error;
+}
+
+otError UniqueServiceManager::AddNetworkDataLocalService(const otServiceConfig &aConfig)
+{
+    return GetNetif().GetNetworkDataLocal().AddService(
+        aConfig.mEnterpriseNumber, aConfig.mServiceData, aConfig.mServiceDataLength, aConfig.mServerConfig.mStable,
+        aConfig.mServerConfig.mServerData, aConfig.mServerConfig.mServerDataLength);
+}
+
+otError UniqueServiceManager::RemoveNetworkDataLocalService(const otServiceConfig &aConfig)
+{
+    return GetNetif().GetNetworkDataLocal().RemoveService(aConfig.mEnterpriseNumber, aConfig.mServiceData,
+                                                          aConfig.mServiceDataLength);
+}
+
+bool UniqueServiceManager::DefaultServerCompare(const otServerConfig *aServerA,
+                                                const otServerConfig *aServerB,
+                                                void *                aContext)
+{
+    bool rval = false;
+
+    if ((Mle::Mle::IsActiveRouter(aServerA->mRloc16) && Mle::Mle::IsActiveRouter(aServerB->mRloc16)) ||
+        (!Mle::Mle::IsActiveRouter(aServerA->mRloc16) && !Mle::Mle::IsActiveRouter(aServerB->mRloc16)))
+    {
+        if (aServerA->mRloc16 <= aServerB->mRloc16)
+        {
+            rval = true;
+        }
+        else
+        {
+            rval = false;
+        }
+    }
+    else if (Mle::Mle::IsActiveRouter(aServerA->mRloc16))
+    {
+        rval = true;
+    }
+    else if (Mle::Mle::IsActiveRouter(aServerA->mRloc16))
+    {
+        rval = false;
+    }
```

Comment:
How about updating this block like below since the default rval is false?
```
if (Mle::Mle::IsActiveRouter(aServerA->mRloc16) == Mle::Mle::IsActiveRouter(aServerB->mRloc16)
{
      if ( aServerA->mRloc16 <= aServerB->mRloc16)
     {
            rval = true;
     }
}
else if (Mle::Mle::IsActiveRouter(aServerA->mRloc16))
{
      rval = true;
}
```

Reply:
Great. I have updated.